### PR TITLE
New resource - azurerm_data_factory_integration_runtime_self_hosted

### PR DESCRIPTION
**Pull request by pearcec**
This resource type supports but a standalone Self-hosted and linking via rbac_authorization. The
key_authorization has no documentation other than the REST interface.  The Powershell Cmdlet does
not make use of it.  So it has been left out.

A full test is setup and necessary for the linked environment to work.

An example and documetnation is included to make the linked setup easier to understand.  People
may also reference the test case.

Fixing issue #6503.

### CODE REVIEW

**Comment by sdebruyn** [comment]
I really would like to have this resource. Is there anything I can do to help? I see that it has a `documentation` label? Does that mean that it needs docs before it can go through? Maybe a CONTRIBUTING.md file could be useful to let us know what we can do to make it easier for you guys :)

File: azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "azurerm_data_factory_integration_self_hosted" {
+			continue
+		}
+
+		name := rs.Primary.Attributes["name"]
+		factoryName := rs.Primary.Attributes["data_factory_name"]
+		resourceGroup := rs.Primary.Attributes["resource_group_name"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go:263]: could we use a resource specific parse ID function here on the resources's id?

File: azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+		// Ensure we have enough information in state to look up in API
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		name := rs.Primary.Attributes["name"]
+		factoryName := rs.Primary.Attributes["data_factory_name"]
+		resourceGroup, hasResourceGroup := rs.Primary.Attributes["resource_group_name"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go:234]: could we use a resource specific parseID function here?

File: azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func testAccAzureRMDataFactoryIntegrationRuntimeSelfHosted_basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go:61]: could we use
```suggestion
  name     = "acctestRG-df-%d"
```

File: azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  type                 = "CustomScriptExtension"
+  type_handler_version = "1.10"
+  settings = jsonencode({
+    "fileUris"         = ["https://raw.githubusercontent.com/Azure/azure-quickstart-templates/00b79d2102c88b56502a63041936ef4dd62cf725/101-vms-with-selfhost-integration-runtime/gatewayInstall.ps1"],
+    "commandToExecute" = "powershell -ExecutionPolicy Unrestricted -File gatewayInstall.ps1 ${azurerm_data_factory_integration_runtime_self_hosted.host.auth_key_1} && timeout /t 120"
+  })
+}
+
+resource "azurerm_resource_group" "host" {
+  name     = "acctesthostRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go:170]: and 
```suggestion
  name     = "acctestRG-df-%d-host"
```

File: azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+resource "azurerm_data_factory" "test" {
+  name                = "acctestdfirsh%d"
+  location            = "${azurerm_resource_group.test.location}"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+}
+
+resource "azurerm_data_factory_integration_runtime_self_hosted" "test" {
+  name                = "self-integration-runtime"
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/resource_arm_data_factory_integration_runtime_self_hosted_test.go:72]: could we use acctest here
```suggestion
  name                = "acctestSIR%d"
```

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	factoryName := id.Path["factories"]
+	name := id.Path["integrationruntimes"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:225]: could we write a parse ID function specific to this resource and use it here like in #5692

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	factoryName := id.Path["factories"]
+	name := id.Path["integrationruntimes"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:159]: and a resource specific parse ID function here

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+	description := d.Get("description").(string)
+
+	selfHostedIntegrationRuntime := datafactory.SelfHostedIntegrationRuntime{
+		Description: &description,
+		Type:        datafactory.TypeSelfHosted,
+		SelfHostedIntegrationRuntimeTypeProperties: expandAzureRmDataFactoryIntegrationRuntimeSelfHostedTypeProperties(d),
+	}
+
+	basicIntegrationRuntime, _ := selfHostedIntegrationRuntime.AsBasicIntegrationRuntime()
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:123]: we should be checking the result of the cast for nil

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(
+					regexp.MustCompile(`^[A-Za-z0-9]+(?:-[A-Za-z0-9]+)*$`),
+					`Invalid name for Self-Hosted Integration Runtime: minimum 3 characters, must start and end with a number or a letter, may only consist of letters, numbers and dashes and no consecutive dashes.`,
+				),
+			},
+
+			"description": {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:50]: we should place optional properties after required

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+				ForceNew:     true,
+				ValidateFunc: validate.DataFactoryName(),
+			},
+
+			"auth_key_1": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"auth_key_2": {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:67]: computed properties should go after optional

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go (hunk 7 above)
Comment (pearcec) [reply azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go:159]: None of these are IDs they are names.  So there isn't really any path to parse.  Both name and data_factory_name are validated via the schema.

**Comment by pearcec** [comment]
@katbyte Hi - I made some adjustments and committed the changes.  There are also some inline comments to the ParseID request.

**Comment by Stanpol** [comment]
Can we expect that this PR will be a part of v2.15.0 milestone?

**Comment by pearcec** [comment]
I updated the code to address @katbyte changes.

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go (hunk 7 above)
Comment (jackofallops) [reply azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go]: Hi @pearcec 
I think Kt was referring to the pattern of a custom parser for the whole resource ID (e.g. `/subscriptions/12345678-1234-1234-1234-12345678abc/resourceGroups/exampleResourceGroup/providers/Microsoft.DataFactory/factories/exampleFactoryName/integrationruntimes/exampleIntegrationRuntime` which could be parsed into a custom type, and then accessed as `id.FactoryName` and `id.Name` - (example [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/compute/parse/image.go))

Comment (pearcec) [reply azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go]: ok makes sense, I will take a look.

File: azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func testAccAzureRMDataFactoryIntegrationRuntimeSelfHosted_rbac(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go:86]:
```suggestion
  name     = "acctestRG-df-%d"
```

File: azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+	ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "azurerm_data_factory_integration_self_hosted" {
+			continue
+		}
+
+		name := rs.Primary.Attributes["name"]
+		factoryName := rs.Primary.Attributes["data_factory_name"]
+		resourceGroup := rs.Primary.Attributes["resource_group_name"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go:263]: Could we use a resource specific parse id function here like can be seen in the analysis package: parse.AnalysisServicesServerID

File: azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+
+		// Ensure we have enough information in state to look up in API
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		name := rs.Primary.Attributes["name"]
+		factoryName := rs.Primary.Attributes["data_factory_name"]
+		resourceGroup, hasResourceGroup := rs.Primary.Attributes["resource_group_name"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/tests/data_factory_integration_runtime_self_hosted_test.go:234]: Could we use a resource specific parse id function here like can be seen in the analysis package: parse.AnalysisServicesServerID

File: azurerm/internal/services/datafactory/data_factory_integration_runtime_self_hosted.go [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	factoryName := id.Path["factories"]
+	name := id.Path["integrationruntimes"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_integration_runtime_self_hosted.go:228]: Could we use a resource specific parse id function here like can be seen in the analysis package: parse.AnalysisServicesServerID

File: azurerm/internal/services/datafactory/data_factory_integration_runtime_self_hosted.go [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	factoryName := id.Path["factories"]
+	name := id.Path["integrationruntimes"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_integration_runtime_self_hosted.go:162]: Could we use a resource specific parse id function here like can be seen in the analysis package: parse.AnalysisServicesServerID

File: azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go (hunk 7 above)
Comment (pearcec) [reply azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go]: Ok I added it.

**Pull request merged**